### ISSUE TITLE
Focus when selecting an item inside a Multiple Selection in a Control that currently does not have Focus does not change Focus as Expected

### ISSUE THREAD

**Issue by cmiles** [punker76/gong-wpf-dragdrop#235]
## What steps will reproduce this issue?

In the Showcase - 1st View ListBox|Bound:

1. Select Item 1 thru Item 5 in the left ListBox and Drag/Move them to the Right ListBox
2. In the left ListBox left click Item 6 and shift left click Item 8 so that the selection is Item 6-Item 8
3. In the right ListBox left click Item 1 (focus will move to Item 1 in the right ListBox as expected)
4. Left click Item 7 in the left ListBox - the selection in the left ListBox will be correct (single no modifiers left click on Item 7 causes Item 7 to be the only item selected) but the focus remains on Item 1 in the right ListBox.
5. To see the unexpected results here hit Control-A - because focus remained on Item 1 in the right ListBox all items in the right ListBox are now selected, but I believe a user would expect Item 7 to have focus since they just clicked it so the results from hitting Control-A are unexpected/incorrect.

### Expected outcome

Clicking an item in a DragDrop enabled control that doesn't have focus should result in the focus moving to that control/item as it would if the control was not Drag and Drop enabled.

I have not created a repo or pull request largely because I am not confident about how to solve this without causing other issues and because in the time I had I didn't understand how to solve this problem generically for a MultiSelector.

Here is the code I am experimenting with in ItemsControlExtensions.cs SetSelectedItem(this ItemsControl itemsControl, object item) - so far this has corrected the problem in the Showcase and in my own app but is so lightly tested that I have no idea if it is 'safe' and also dislike that I had to add an else if block for the DataGrid...

The snippet below needs using System.Windows.Input; for the TraversalRequest.

```
public static void SetSelectedItem(this ItemsControl itemsControl, object item)
    {
      if (itemsControl is ListBox)
      {
        var selectionMode = ((ListBox)itemsControl).SelectionMode;
        try
        {
          // change SelectionMode for UpdateAnchorAndActionItem
          ((ListBox)itemsControl).SelectionMode = SelectionMode.Single;
          ((ListBox)itemsControl).SelectedItem = null;
          ((ListBox)itemsControl).SelectedItem = item;

          ((ListBox)itemsControl).UpdateLayout();

          var listBoxItem = (ListBoxItem)((ListBox)itemsControl)
              .ItemContainerGenerator
              .ContainerFromItem(item);

          listBoxItem.Focus();
        }
        finally
        {
          ((ListBox)itemsControl).SelectionMode = selectionMode;
        }
      }
      else if (itemsControl is DataGrid)
      {
        var dataGrid = (DataGrid)itemsControl;

        dataGrid.SelectedItems.Clear();
        dataGrid.SelectedItem = item;

        DataGridRow row = dataGrid.ItemContainerGenerator.ContainerFromItem(item) as DataGridRow;
        if (row == null)
        {
          dataGrid.ScrollIntoView(item);
          row = dataGrid.ItemContainerGenerator.ContainerFromItem(item) as DataGridRow;
        }
        row?.MoveFocus(new TraversalRequest(FocusNavigationDirection.Next));
      }
      else if (itemsControl is TreeView)
      {
        // clear old selected item
        var prevSelectedItem = ((TreeView)itemsControl).GetValue(TreeView.SelectedItemProperty);
        if (prevSelectedItem != null)
        {
          var prevSelectedTreeViewItem =
              ((TreeView)itemsControl).ItemContainerGenerator.ContainerFromItem(prevSelectedItem) as TreeViewItem;
          if (prevSelectedTreeViewItem != null)
          {
            prevSelectedTreeViewItem.IsSelected = false;
          }
        }
        // set new selected item
        // TreeView.SelectedItemProperty is a read only property, so we must set the selection on the TreeViewItem itself
        var treeViewItem = ((TreeView)itemsControl).ItemContainerGenerator.ContainerFromItem(item) as TreeViewItem;
        if (treeViewItem != null)
        {
          treeViewItem.IsSelected = true;
        }
      }
      else if (itemsControl is MultiSelector)
      {
        ((MultiSelector)itemsControl).SelectedItem = null;
        ((MultiSelector)itemsControl).SelectedItem = item;
      }
      else if (itemsControl is Selector)
      {
        ((Selector)itemsControl).SelectedItem = null;
        ((Selector)itemsControl).SelectedItem = item;
      }
    }
```

### Environment

- GongSolutions.WPF.DragDrop __Checked out the 2/6/2017 Commit 3c87996 version on 2/25/2017__
- Windows OS __Version Win 10 Pro - 10.0.14393 Build 14393__
- Visual Studio __2017__
- .NET Framework __Tested current Showcase which targets 4.5__

**Comment by airbreather** [comment]
It didn't get automatically linked for some reason... #239 looks like it would fix this, if merged.

**Comment by punker76** [comment]
Closed by #239
